File: kanvas/src/vkMain/kotlin/com/cws/kanvas/rendering/backend/VulkanAllocator.kt
package com.cws.kanvas.rendering.backend

import com.cws.print.Print
import org.lwjgl.system.MemoryUtil.*
import org.lwjgl.util.vma.Vma
import org.lwjgl.util.vma.VmaAllocatorCreateInfo
import org.lwjgl.vulkan.VkAllocationCallbacks

object VulkanAllocator : Resource<Long, Unit>(Unit) {

    private const val TAG = "VulkanAllocator"

    var context: RenderContext? = null

    val callbacks: VkAllocationCallbacks =
        VkAllocationCallbacks.create()
            .pUserData(0)
            .pfnAllocation(::vkAllocate)
            .pfnFree(::vkFree)
            .pfnReallocation(::vkReallocate)
            .pfnInternalAllocation(::vkAllocateNotification)
            .pfnInternalFree(::vkFreeNotification)

    init {
        callbacks.pUserData(thisPtr())
    }

    override fun onCreate() {
        val instance = context?.handle ?: return
        val physicalDevice = context?.device?.physicalDevice ?: return
        val device = context?.device?.handle ?: return

        handle = stack { stack ->
            val createInfo = VmaAllocatorCreateInfo.calloc()
                .instance(instance)
                .physicalDevice(physicalDevice)
                .device(device)
                .pAllocationCallbacks(callbacks)
            val pAllocator = stack.mallocPointer(1)
            Vma.vmaCreateAllocator(createInfo, pAllocator)
            pAllocator.get()
        }
    }

    override fun onRelease() {
        handle?.let { handle ->
            Vma.vmaDestroyAllocator(handle)
        }
        handle = null
    }

    private fun allocate(
        size: Long,
        alignment: Long,
        scope: Int
    ): Long {
        Print.d(TAG, "allocate: size=$size alignment=$alignment scope=$scope")

        val align = maxOf(alignment, Long.SIZE_BYTES.toLong())
        val totalSize = size + align - 1 + Long.SIZE_BYTES

        val original = nmemAlloc(totalSize)
        if (original == 0L) return 0L

        val raw = original + Long.SIZE_BYTES
        val aligned = (raw + align - 1) and (align - 1).inv()

        memPutAddress(aligned - Long.SIZE_BYTES, original)

        return aligned
    }

    private fun free(address: Long) {
        Print.d(TAG, "free: ${address.toString(16)}")
        if (address != 0L) {
            val original = memGetAddress(address - Long.SIZE_BYTES)
            nmemFree(original)
        }
    }

    private fun reallocate(
        oldAddress: Long,
        size: Long,
        alignment: Long,
        scope: Int
    ): Long {
        Print.d(TAG, "reallocate: size=$size alignment=$alignment scope=$scope")

        val newAddress = allocate(size, alignment, scope)
        if (oldAddress != 0L && newAddress != 0L) {
            memCopy(oldAddress, newAddress, size)
            free(oldAddress)
        }
        return newAddress
    }

    private fun allocateNotification(
        size: Long,
        type: Int,
        scope: Int
    ) {
        Print.d(TAG, "allocateNotification: size=$size type=$type scope=$scope")
    }

    private fun freeNotification(
        size: Long,
        type: Int,
        scope: Int
    ) {
        Print.d(TAG, "freeNotification: size=$size type=$type scope=$scope")
    }

    private fun thisPtr(): Long = System.identityHashCode(this).toLong()

    private fun vkAllocate(
        userData: Long,
        size: Long,
        alignment: Long,
        scope: Int
    ): Long {
        return allocate(size, alignment, scope)
    }

    private fun vkFree(
        userData: Long,
        memory: Long
    ) {
        free(memory)
    }

    private fun vkReallocate(
        userData: Long,
        oldMemory: Long,
        size: Long,
        alignment: Long,
        scope: Int
    ): Long {
        return reallocate(oldMemory, size, alignment, scope)
    }

    private fun vkAllocateNotification(
        userData: Long,
        size: Long,
        type: Int,
        scope: Int
    ) {
        allocateNotification(size, type, scope)
    }

    private fun vkFreeNotification(
        userData: Long,
        size: Long,
        type: Int,
        scope: Int
    ) {
        freeNotification(size, type, scope)
    }
}

File: kanvas-vk/src/cpp/vk/api/VulkanAllocator.cpp
//
// Created by cheerwizard on 18.10.25.
//

#include "VulkanAllocator.hpp"

#define VMA_IMPLEMENTATION
#include "vk_mem_alloc.h"

#include "../core/logger.hpp"

VKAPI_ATTR void* VKAPI_CALL vk_allocate(void* userData, size_t size, size_t alignment, VkSystemAllocationScope scope) {
    return static_cast<VulkanAllocator*>(userData)->allocate(size, alignment, scope);
}

VKAPI_ATTR void VKAPI_CALL vk_free(void* userData, void* memory) {
    static_cast<VulkanAllocator*>(userData)->free(memory);
}

VKAPI_ATTR void* VKAPI_CALL vk_reallocate(void* userData, void* oldMemory, size_t size, size_t alignment, VkSystemAllocationScope scope) {
    return static_cast<VulkanAllocator*>(userData)->reallocate(oldMemory, size, alignment, scope);
}

VKAPI_ATTR void VKAPI_CALL vk_allocate_notification(void* userData, size_t size, VkInternalAllocationType type, VkSystemAllocationScope scope) {
    static_cast<VulkanAllocator*>(userData)->allocateNotification(size, type, scope);
}

VKAPI_ATTR void VKAPI_CALL vk_free_notification(void* userData, size_t size, VkInternalAllocationType type, VkSystemAllocationScope scope) {
    static_cast<VulkanAllocator*>(userData)->freeNotification(size, type, scope);
}

VulkanAllocator::VulkanAllocator() {
    callbacks.pUserData = this;
    callbacks.pfnAllocation = vk_allocate;
    callbacks.pfnFree = vk_free;
    callbacks.pfnReallocation = vk_reallocate;
    callbacks.pfnInternalAllocation = vk_allocate_notification;
    callbacks.pfnInternalFree = vk_free_notification;
}

VulkanAllocator& VulkanAllocator::getInstance() {
    static VulkanAllocator instance;
    return instance;
}

void VulkanAllocator::New(VkInstance instance, VkPhysicalDevice physicalDevice, VkDevice device) {
    VmaAllocatorCreateInfo createInfo = {
        .physicalDevice = physicalDevice,
        .device = device,
        .pAllocationCallbacks = &callbacks,
        .instance = instance
    };
    vmaCreateAllocator(&createInfo, &allocator);
}

void VulkanAllocator::Delete() const {
    vmaDestroyAllocator(allocator);
}

void* VulkanAllocator::allocate(size_t size, size_t alignment, VkSystemAllocationScope scope) {
    LOG_INFO(TAG, "allocate: size=%d, alignment=%d, scope=%d", size, alignment, scope);

    if (alignment < sizeof(void*)) {
        alignment = sizeof(void*);
    }

    // Overallocate: extra space for alignment + to store original pointer
    size_t total_size = size + alignment - 1 + sizeof(void*);
    void *original = malloc(total_size);
    if (!original) return nullptr;

    // Calculate aligned address within the allocated block
    uintptr_t raw = (uintptr_t)original + sizeof(void*);
    uintptr_t aligned = (raw + alignment - 1) & ~(uintptr_t)(alignment - 1);

    // Store original pointer just before aligned pointer
    ((void**)aligned)[-1] = original;

    return (void*) aligned;
}

void VulkanAllocator::free(void *address) {
    LOG_INFO(TAG, "free: %p", address);
    if (address) {
        // Retrieve original pointer and free
        ::free(((void**)address)[-1]);
    }
}

void* VulkanAllocator::VulkanAllocator::reallocate(void *oldAddress, size_t size, size_t alignment, VkSystemAllocationScope scope) {
    LOG_INFO(TAG, "reallocate: size=%d, alignment=%d, scope=%d", size, alignment, scope);
    void* newAddress = allocate(size, alignment, scope);
    memcpy(newAddress, oldAddress, size);
    free(oldAddress);
    return newAddress;
}

void VulkanAllocator::allocateNotification(size_t size, VkInternalAllocationType type, VkSystemAllocationScope scope) {
    LOG_INFO(TAG, "allocateNotification: size=%d, type=%d, scope=%d", size, type, scope);
}

void VulkanAllocator::freeNotification(size_t size, VkInternalAllocationType type, VkSystemAllocationScope scope) {
    LOG_INFO(TAG, "freeNotification: size=%d, type=%d, scope=%d", size, type, scope);
}